### PR TITLE
Allow unstable items to be re-exported unstably without requiring the feature be enabled

### LINKED ISSUE #94972: Allow `std` to re-export unstable things from `core` without having to enable the feature itself
`library/std/src/lib.rs` now enables a lot of `#![feature]`s from `core` and `alloc` and `test` only because it needs to re-export some unstable items (with the same `#[unstable]` tag). It makes it hard to tell if `std` needs those features itself or not.

It'd be great if an `#[unstable] pub use` would work without enabling the feature in the entire crate. (As long as the feature name is the same as on the original item, I suppose.)

These are all not directly used by `std`, but only used for items that are unstably re-exported:

```
#![feature(assert_matches)]
#![feature(async_iterator)]
#![feature(c_variadic)]
#![feature(cfg_accessible)]
#![feature(cfg_eval)]
#![feature(concat_bytes)]
#![feature(const_format_args)]
#![feature(c_size_t)]
#![feature(core_ffi_c)]
#![feature(core_panic)]
#![feature(custom_test_frameworks)]
#![feature(edition_panic)]
#![feature(format_args_nl)]
#![feature(log_syntax)]
#![feature(once_cell)]
#![feature(saturating_int_impl)]
#![feature(stdsimd)]
#![feature(test)]
#![feature(trace_macros)]
```

I'd love to remove these lines.

### PR DESCRIPTION
**Pull request by semicoleon**
Closes #94972

The diagnostic may need some work still, and I haven't added a test yet

### CODE REVIEW

**Comment by lcnr** [comment]
don't trust myself to review anything loosing our stability checks

r? @petrochenkov

File: compiler/rustc_passes/src/stability.rs [hunk 1]
```diff
@@ -805,14 +806,69 @@ impl<'tcx> Visitor<'tcx> for Checker<'tcx> {
...
+    tcx: TyCtxt<'tcx>,
+    path: &'a hir::Path<'tcx>,
+    id: hir::HirId,
+) -> bool {
+    // Get the LocalDefId so we can lookup the item to check the kind.
+    let def_id = if let Some(def_id) = tcx.hir().opt_local_def_id(id) {
+        def_id
+    } else {
+        return false;
+    };
```
Comment (petrochenkov) [inline compiler/rustc_passes/src/stability.rs:830]: Code here and below could use the `let ... else` feature.
```rust
let Some(def_id) = tcx.hir().opt_local_def_id(id) else {
    return false;
}
```

File: compiler/rustc_passes/src/stability.rs [hunk 2]
```diff
@@ -805,14 +806,69 @@ impl<'tcx> Visitor<'tcx> for Checker<'tcx> {
...
+        stab
+    } else {
+        return false;
+    };
+
+    let item = tcx.hir().item(hir::ItemId { def_id });
+    let item_kind = &item.kind;
+
+    // If this is a path that isn't a use, we don't need to do anything special
+    if !matches!(item_kind, ItemKind::Use(..)) {
```
Comment (petrochenkov) [inline compiler/rustc_passes/src/stability.rs:842]:
```suggestion
    if !matches!(tcx.hir().item(hir::ItemId { def_id }).kind, ItemKind::Use(..)) {
```

File: compiler/rustc_passes/src/stability.rs [hunk 3]
```diff
@@ -805,14 +806,69 @@ impl<'tcx> Visitor<'tcx> for Checker<'tcx> {
...
+        let method_span = path.segments.last().map(|s| s.ident.span);
+
+        tcx.eval_stability(def_id, Some(id), path.span, method_span)
+    } else {
+        // If the path doesn't resolve to anything, there's nothing more to check.
+        return false;
+    };
+
+    // If we are going to deny the usage, check whether it's a re-export first with the same unstable feature.
+    if let rustc_middle::middle::stability::EvalResult::Deny { feature, .. } = decision {
```
Comment (petrochenkov) [inline compiler/rustc_passes/src/stability.rs:856]: I don't think this condition is necessary?
It doesn't matter whether the reexport target is unstable or not, and whether the feature is enabled or disabled, the reexport gating can be skipped anyway.

File: compiler/rustc_passes/src/stability.rs [hunk 4]
```diff
@@ -805,14 +806,69 @@ impl<'tcx> Visitor<'tcx> for Checker<'tcx> {
...
+        if stab.level.is_unstable() && feature == stab.feature {
+            // This is a re-export, marked as unstable, with the same feature name. Don't require a `#![feature]` gate
+            return true;
+        } else {
+            // We almost met the requirements but the feature name didn't match. Emit a warning in addition to the standard feature gate error.
+            tcx.sess.parse_sess.struct_warn("Re-exporting an unstable item without a feature gate requires marking the re-export as unstable with the same feature name")
+                // should probably be getting the span of the unstable attr here? Maybe make it an actual suggestion.
+                .span_help(path.span, format!("Use `#![unstable(feature = {})]` to re-export", feature))
+                .emit();
```
Comment (petrochenkov) [inline compiler/rustc_passes/src/stability.rs:865]: Hard-coded warnings should almost never be reported, and this warning is not generally useful.
If the reexport is gated by a separate feature then there are probably reasons for that.

**Comment by petrochenkov** [comment]
This also needs a test.

File: compiler/rustc_passes/src/stability.rs (hunk 3 above)
Comment (semicoleon) [reply compiler/rustc_passes/src/stability.rs]: I went ahead and removed these checks. 

As it is now this PR would completely prevent any `use` items from triggering a "use of unstable" error, only the actual usages of the item trigger errors. Do you think checking whether staged_api is enabled would be good?

**Comment by semicoleon** [comment]
When I was playing around with the test, I noticed that I could get master to pass a version of the test that looks like it should probably fail. I pushed that version of the test here https://github.com/semicoleon/rust/commit/82b7fc22d2f1df41042b46b66905f12a13634ca2

I think we're just not checking the stability attributes on re-exports at all. So if you enable the feature the actual resolved item requires, any `#[unstable(...)]` attributes on re-exports between the resolved item and the call site don't produce an error the way I'd expect it to. 

I'm not sure that's _particularly_ relevant to this PR, or whether it's really worth worrying about, but I figured I'd mention it

**Comment by semicoleon** [comment]
I went ahead and implemented a check for "public" visibility since the test failures were all on non-public `use`s. I also moved the check for not emitting unstable errors inside eval_stability so the deprecation warnings would still run in all cases.

**Comment by semicoleon** [comment]
That last test failure is a re-export, so we might need to change that test if this version of the implementation is what we want to go with

[This is the test that's failing](https://github.com/rust-lang/rust/blob/master/src/test/ui-fulldeps/pathless-extern-unstable.rs)

File: compiler/rustc_passes/src/stability.rs [hunk 5]
```diff
@@ -816,48 +816,20 @@ impl<'tcx> Visitor<'tcx> for Checker<'tcx> {
...
-        return false;
-    };
-
-    let method_span = path.segments.last().map(|s| s.ident.span);
-
-    let decision = tcx.eval_stability(def_id, Some(id), path.span, method_span);
-
-    // If we are going to deny the usage, check whether it's a re-export first with the same unstable feature.
-    if let rustc_middle::middle::stability::EvalResult::Deny { feature, .. } = decision {
-        if stab.level.is_unstable() && feature == stab.feature {
```
Comment (petrochenkov) [inline compiler/rustc_passes/src/stability.rs]: This check was necessary, it ensured that the `use` item itself had an `#[unstable]` attribute.
Without it any third-party crate can reexport something unstable from the standard library on stable compiler.

**Comment by petrochenkov** [comment]
>so the deprecation warnings would still run in all cases

Yeah, that's probably better, at least for a start.
If people are not asking to skip deprecation messages in cases like `#[unstable] pub use something_deprecated;`, then let's keep them.
Ping @m-ou-se just in case.

File: compiler/rustc_passes/src/stability.rs [hunk 6]
```diff
@@ -807,12 +808,44 @@ impl<'tcx> Visitor<'tcx> for Checker<'tcx> {
...
+            )
         }
         intravisit::walk_path(self, path)
     }
 }
 
+/// Check whether a path is a `use` item that has been marked as unstable.
+fn is_unstable_reexport<'tcx>(tcx: TyCtxt<'tcx>, id: hir::HirId) -> bool {
+    // Get the LocalDefId so we can lookup the item to check the kind.
+    let Some(def_id) = tcx.hir().opt_local_def_id(id) else { return false};
```
Comment (fbstj) [inline compiler/rustc_passes/src/stability.rs:830]:
```suggestion
    let Some(def_id) = tcx.hir().opt_local_def_id(id) else { return false; };
```
is that a bug in rustfmt?

Comment (lcnr) [reply compiler/rustc_passes/src/stability.rs:830]: rustfmt doesn't yet format `let else`. As rustfmt changing the formatting is only done once they've decided on a stable formatting style

File: compiler/rustc_passes/src/stability.rs [hunk 7]
```diff
@@ -807,12 +808,44 @@ impl<'tcx> Visitor<'tcx> for Checker<'tcx> {
...
+                } else {
+                    AllowUnstable::No
+                },
+            )
         }
         intravisit::walk_path(self, path)
     }
 }
 
+/// Check whether a path is a `use` item that has been marked as unstable.
```
Comment (petrochenkov) [inline compiler/rustc_passes/src/stability.rs:827]: Could you link to #94972 to give some explanation for this special case?

**Comment by petrochenkov** [comment]
LGTM, could you also squash commits after addressing the remaining comments?
@rustbot author

**Comment by RalfJung** [comment]
This seems to be the likely cause of a subtle bug, but more investigation is definitely needed: https://github.com/rust-lang/rust/issues/114838.

**Comment by RalfJung** [comment]
Ah: this patch also kicks in when `-Zforce-unstable-if-unmarked` is enabled. That's not intentional, is it?

**Comment by semicoleon** [comment]
I don't believe so, no

**Comment by RalfJung** [comment]
The problem is that it kicks in for every unstable reexport -- whether it has an explicit `#[unstable]` or whether it is implicitly unstable due to the `-Z` flag.

Maybe it should be gated on "the item is unstable and the crate explicitly enabled `staged_api`?

**Comment by petrochenkov** [comment]
For the purpose of fixing https://github.com/rust-lang/rust/issues/94972, I think it would be enough if `AllowUnstable::Yes` was enabled only for reexports with explicit `unstable` attributes.

**Comment by RalfJung** [comment]
Or that, yes. (Then in std it wouldn't kick in when `unstable` is inherited from the module.)

**Pull request merged**